File: src/main/java/twopointers/palindrome/AlmostPalindrome.java
package twopointers.palindrome;

public class AlmostPalindrome {

    public boolean isPalindrome(String s) {

        var low = 0;
        var high = s.length() - 1;
        var canContinue = true;
        while (low < high) {
            if (s.charAt(low) != s.charAt(high)) {
                if (canContinue) {
                    if (low == high - 1) return true;
                    canContinue = false;
                    if (s.charAt(low + 1) == s.charAt(high)) low++;
                    else if (s.charAt(low) == s.charAt(high - 1)) high--;
                    else return false;
                } else return false;
            }
            low++;
            high--;
        }
        return true;
    }

}

File: src/main/kotlin/twopointers/palindrome/AlmostPalindrome.kt
package twopointers.palindrome

fun isAlmostPalindrome(s: String): Boolean {
    var low = 0
    var high = s.lastIndex
    var canContinue = true
    while (low < high) {
        if (s[low] != s[high]) {
            if (canContinue) {
                if (low == high - 1) return true
                canContinue = false
                if (s[low + 1] == s[high]) low++
                else if (s[low] == s[high - 1]) high--
                else return false
            } else return false
        }
        low++
        high--
    }
    return true
}
